### PR TITLE
Keep features after transform

### PR DESCRIPTION
**Pull request by lhoestq**
When applying a transform like `map`, some features were lost (and inferred features were used).
It was the case for ClassLabel, Translation, etc.

To fix that, I did some modifications in the `ArrowWriter`:

- added the `update_features` parameter. When it's `True`, then the features specified by the user (if any) can be updated with inferred features if their type don't match. `map` transform sets `update_features=True` when writing to cache file or buffer. Features won't change by default in `map`.

- added the `with_metadata` parameter. If `True`, the `features` (after update) will be written inside the metadata of the schema in this format:
```
{
    "huggingface": {"features" : <serialized Features exactly like dataset_info.json>}
} 
```
Then, once a dataset is instantiated without info/features, these metadata are used to set the features of the dataset.

### CODE REVIEW

File: src/nlp/features.py [hunk 1]
```diff
@@ -277,7 +277,7 @@ class Translation:
     _type: str = field(default="Translation", init=False, repr=False)
 
     def __call__(self):
-        return pa.struct({lang: pa.string() for lang in self.languages})
+        return pa.struct({lang: pa.string() for lang in sorted(self.languages)})
```
Comment (lhoestq) [inline src/nlp/features.py:280]: I had to add `sorted` in some places to make features comparisons work (features.type needs to be deterministic)

File: tests/test_arrow_dataset.py [hunk 2]
```diff
@@ -309,6 +323,82 @@ def test_filter(self):
...
+        dset = Dataset.from_dict({"tokens": [["foo"] * 5] * 10, "labels": [[1] * 5] * 10}, features=features)
+
+        def invert_labels(x):
+            return {"labels": [(1 - label) for label in x["labels"]]}
+
+        inverted_dset = dset.map(invert_labels, keep_in_memory=True)
+        self.assertEqual(inverted_dset.features.type, features.type)
+        self.assertDictEqual(inverted_dset.features, features)
+
+    def test_keep_features_after_loading_from_cache(self):
```
Comment (thomwolf) [inline tests/test_arrow_dataset.py:385]: Can you add a last test about when you add a new column?
E.g. create a duplicate `labels2` of the `labels` column and check that it's assigned an inferred features (should be `Sequence(Value('int32'))` I guess?) while the original `labels` column still has a `Sequence(ClassLabel(names=["negative", "positive"]))` feature

Comment (lhoestq) [reply tests/test_arrow_dataset.py:385]: int64 actually ^^

File: tests/test_arrow_dataset.py [hunk 3]
```diff
@@ -309,6 +323,105 @@ def test_filter(self):
...
+                "labels": Sequence(ClassLabel(names=["negative", "positive"])),
+                "labels2": Sequence(Value("int64")),
+            }
+        )
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            tmp_file = os.path.join(tmp_dir, "test.arrow")
+            inverted_dset = dset.map(invert_labels, cache_file_name=tmp_file)
+            self.assertEqual(inverted_dset.features.type, expected_features.type)
+            self.assertDictEqual(inverted_dset.features, expected_features)
```
Comment (lhoestq) [inline tests/test_arrow_dataset.py:423]: I just added this test with a new column to check that inference is fine @thomwolf

File: src/nlp/features.py [hunk 4]
```diff
@@ -458,7 +460,10 @@ def generate_from_arrow_type(pa_type: pa.DataType):
     elif isinstance(pa_type, pa.FixedSizeListType):
         return Sequence(feature=generate_from_arrow_type(pa_type.value_type), length=pa_type.list_size)
     elif isinstance(pa_type, pa.ListType):
-        return [generate_from_arrow_type(pa_type.value_type)]
+        feature = generate_from_arrow_type(pa_type.value_type)
+        if isinstance(feature, (dict, tuple, list)):
+            return [feature]
+        return Sequence(feature=feature)
```
Comment (lhoestq) [inline src/nlp/features.py:466]: I had to fix that.
There were issues with arrow types like `list of list of struct`.
We have one dataset that has this format: `docred`

Comment (thomwolf) [reply src/nlp/features.py:466]: Great catch!

**Comment by lhoestq** [comment]
One note on features inference:

if an arrow type is `struct of items` where each item is a `list`, then we return a `dict` in which each item is a `Sequence`.
It means that we don't use the Sequence <-> dict swap when we infer features.

It's fine because the swap is generally used in dataset scripts, in which features are defined (inferred features are discarded)

**Comment by lhoestq** [comment]
If it's fine for you @thomwolf we can merge this one :)

**Comment by thomwolf** [comment]
Yes this is fine I think!

**Pull request merged**